File: 15. Edit Distance (Levenshtein Distance).py
# 15. Edit Distance (Levenshtein Distance)
# Topic: Edit Distance
# Description: Compute minimum operations (insert/delete/replace) to convert string a into b using memoization.
# Input: two lines: string a and string b
# Output: minimum edit distance

import sys
from functools import lru_cache

if __name__ == '__main__':
    lines = sys.stdin.read().splitlines()
    if not lines:
        sys.exit(0)
    a = lines[0].rstrip('\n')
    b = lines[1].rstrip('\n') if len(lines) > 1 else ''
    n, m = len(a), len(b)

    @lru_cache(maxsize=None)
    def edit(i: int, j: int) -> int:
        if i == n:
            return m - j
        if j == m:
            return n - i
        if a[i] == b[j]:
            return edit(i+1, j+1)
        return 1 + min(edit(i+1, j), edit(i, j+1), edit(i+1, j+1))

    print(edit(0,0))

# Memoization process started at edit decorated with lru_cache.
# Time Complexity: O(n * m)
# Space Complexity: O(n * m)

File: 15. Edit Distance (Levenshtein Distance).cpp
// 15. Edit Distance (Levenshtein Distance)
// Topic: Edit Distance
// Description: Compute minimal insert/delete/replace operations using recursion + memoization.
// Input: two strings a and b (each on separate lines)
// Output: minimum edit distance

#include <bits/stdc++.h>
using namespace std;

int editDist(int i, int j, const string& a, const string& b, vector<vector<int>>& memo)
{
    if (i == (int)a.size())
        return (int)b.size() - j;
    if (j == (int)b.size())
        return (int)a.size() - i;
    if (memo[i][j] != -1)
        return memo[i][j];
    if (a[i] == b[j])
        return memo[i][j] = editDist(i + 1, j + 1, a, b, memo);
    int insertOp = 1 + editDist(i, j + 1, a, b, memo);
    int deleteOp = 1 + editDist(i + 1, j, a, b, memo);
    int replaceOp = 1 + editDist(i + 1, j + 1, a, b, memo);
    return memo[i][j] = min({insertOp, deleteOp, replaceOp});
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    string a, b;
    if (!getline(cin, a))
        return 0;
    if (!getline(cin, b))
        return 0;
    vector<vector<int>> memo(a.size(), vector<int>(b.size(), -1));
    cout << editDist(0, 0, a, b, memo) << '\n';
}

// Memoization process started in editDist where memo[i][j] stores minimal operations for suffixes.
// Time Complexity: O(|a| * |b|)
// Space Complexity: O(|a| * |b|)
